runtime/compiler/runtime/JITServerAOTCache.cpp

```diff
@@ -812,13 +890,70 @@ JITServerAOTCacheMap::get(const std::string &name, uint64_t clientUID)
       return NULL;
       }
 
+   // See if we can load the cache from disk
+   TR::CompilationInfo *compInfo = TR::CompilationInfo::get();
+   if (compInfo->getPersistentInfo()->getJITServerUseAOTCachePersistence() &&     // Persistence feature is enabled
+       _cachesExcludedFromLoading.find(name) == _cachesExcludedFromLoading.end()) // I am allowed to load this cache from disk
+      {
+      if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "AOT cache: See if we can load cache %s from disk\n", name.c_str());
+
+      // Check that we don't have another request in progress for the same named cache
+      if (_cachesBeingLoaded.find(name) != _cachesBeingLoaded.end())
+         {
+         // Another thread is already loading this cache
+         if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "AOT cache: Another thread is already loading cache %s from disk\n", name.c_str());
+         pending = true;
+         return NULL;
+         }
+
+      // Prepend the directory where caches reside and check if my desired cache exists
+      std::string cacheFileName = buildCacheFileName(compInfo->getPersistentInfo()->getJITServerAOTCacheDir(), name);
+      OMRPORT_ACCESS_FROM_OMRPORT(TR::Compiler->omrPortLib);
+      J9FileStat buf;
+      int32_t rc = omrfile_stat(cacheFileName.c_str(), 0, &buf);
```

Comment:
I would consider not `stat`ing the file separately and instead deal with the missing file when we try to open it. This introduces a TOCTTOU issue (a minor one, but still) and adds syscall latency to the current compilation request that bypasses the cache being loaded.

Reply:
When I wrote the code I thought about the scenario where a client asks for a cache name the server has not seen before and therefore there is no cache. If we proceed without verifying the existence of the cache file, the in-memory cache will be created later (when we attempt the read the nonexistent file) and all the methods compiled until then will not be cached. 
I guess only the second client will be affected because the compilations from this one will be cached and the third client will have access to them.

Reply:
I see. Then it does make sense to check if the file exists. Maybe a better solution is to open the file here and pass the file descriptor/stream to the thread that will do the snapshot load.

Reply:
I still think that a better solution is to `fopen()` the file here. If it succeeds, add the `FILE *` along with the name to `_cachesToLoadQueue`. Otherwise the file either doesn't exist or we don't have permissions to read it, and in either case we should create a new cache and not try to read it from the file.

This way we avoid the scenario where cache directory permissions are misconfigured and each JITServer instance will try to read the cache instead of creating a new one, and the first few AOT methods will never be cached. I think that would also simply the implementation a bit. But it's up to you whether you think it's worth it.

Reply:
Thinking some more about this, I am leaning towards eliminating any system call at this point. In kubernetes the snapshot can be far away on NFS or COS (cloud object storage) and may take many milliseconds just to check whether the file exists. This could delay the current compilation (and others because of the cache map lock) a great deal.
The downside of proceeding without a snapshot existence check is that, the first N compilations from a first client that attaches to a JITServer which does not have the requested cache will not be put in the AOT cache. This may happen later on when another client requests the missing methods.

Reply:
One more thought. Once we support merging caches/snapshots, the issue of not caching the first N compilations could be fixed by merging the cache being loaded from the file into the in-memory cache (which we can create and start populating immediately). Assuming that merging doesn't measurably slow down the load.

Reply:
I have eliminated the `stat`-ing of the snapshot. In my experiments I only see one method being compiled before we choose to create an empty in-memory cache.